ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/FollowerPromptness.hs

```diff
@@ -0,0 +1,214 @@
+import           Test.Util.ChainUpdates
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock
+import           Test.Util.Tracer (recordingTracerTVar)
+
+tests :: TestTree
+tests = testGroup "FollowerPromptness"
+    [ testProperty "followerPromptness" prop_followerPromptness
+    ]
+
+prop_followerPromptness :: FollowerPromptnessTestSetup -> Property
+prop_followerPromptness fpts =
+    counterexample ("Trace:\n" <> unlines (ppTrace <$> traceByTime)) $
+    counterexample (condense fpts) $
+    counterexample ("Instruction timings: " <> condense followerInstrTimings) $
+    counterexample ("Failed to pipeline: " <> condense notPipelined)
+    (null notPipelined) .&&.
+    counterexample ("Not processed: " <> condense unprocessed)
+    (null unprocessed)
+  where
+    FollowerPromptnessOutcome{..} =
+      runSimOrThrow $ runFollowerPromptnessTest fpts
+
+    -- Hashes of tentative headers which were not immediately emitted as a
+    -- follower instruction.
+    notPipelined =
+        tentativeHeadersWithoutFollowUp
+          0
+          followerInstrTimings
+
+    -- Hashes of tentative header which were not processed (i.e. made obsolete
+    -- due to adoption, or identified as a trap).
+    unprocessed =
+        tentativeHeadersWithoutFollowUp
+          artificialDelay
+          tentativeHeaderUnsetTimings
+
+    -- Given a collection of timestamped hashes (considered as follow-up events
+    -- of a specific hash), return the timestamped tentative header hashes which
+    -- are not contained therein after the given delay.
+    tentativeHeadersWithoutFollowUp ::
+         DiffTime
+      -> Map Time (Set TestHash)
+      -> [(Time, Set TestHash)]
+    tentativeHeadersWithoutFollowUp delay followUpTimings =
+        [ (time, notFollowedUp)
+        | (time, tentativeHashes) <- Map.toAscList tentativeHeaderSetTimings
+        , let followUpHashes =
+                Map.findWithDefault mempty (addTime delay time) followUpTimings
+              notFollowedUp = tentativeHashes Set.\\ followUpHashes
+        , not $ Set.null notFollowedUp
+        ]
+
+    ppTrace (time, ev) = show time <> ": " <> ev
+
+data FollowerPromptnessOutcome = FollowerPromptnessOutcome {
+    -- | The set of tentative headers by timestamp when set. With the current
+    -- implementation of ChainSel, all sets should contain exactly one element.
+    tentativeHeaderSetTimings   :: Map Time (Set TestHash)
+  , -- | The set of tentative headers by timestamp when set.
+    tentativeHeaderUnsetTimings :: Map Time (Set TestHash)
+  , -- | The set of AddBlock instructions by a tentative follower by timestamp.
+    followerInstrTimings        :: Map Time (Set TestHash)
+  , traceByTime                 :: [(Time, String)]
+  }
+
+runFollowerPromptnessTest ::
+     forall m. IOLike m
+  => FollowerPromptnessTestSetup
+  -> m FollowerPromptnessOutcome
+runFollowerPromptnessTest FollowerPromptnessTestSetup{..} = withRegistry \registry -> do
```

Comment:
Easier to read: `{..} -> {chainUpdates}`

Reply:
As `TestSetup` now has more fields, I kept it like this for now.